src/modm/math/units.hpp

```diff
@@ -0,0 +1,123 @@
+	/// @ingroup modm_architecture_literals
+	/// @{
+	using bitrate_t = uint32_t;
+	template<typename T> constexpr bitrate_t bps(T value);
+	template<typename T> constexpr bitrate_t kbps(T value);
+	template<typename T> constexpr bitrate_t Mbps(T value);
+	/// @}
+
+	/// Converts to Percent
+	/// @ingroup modm_architecture_literals
+	/// @{
+	using percent_t = uint16_t;
+	template<typename T> constexpr percent_t pct(T value);
+	/// @}
+
+	namespace literals
+	{
+		/// Converts to Hertz
+		/// @ingroup modm_architecture_literals
+		/// @{
+		constexpr operator""_Hz(unsigned long long int value);
+		constexpr operator""_kHz(T value);
+		constexpr operator""_MHz(T value);
+		/// @}
+
+		/// Converts to Baud
+		/// @ingroup modm_architecture_literals
+		/// @{
+		constexpr operator""_Bd(T value);
+		constexpr operator""_kBd(T value);
+		constexpr operator""_MBd(T value);
+		/// @}
+
+		/// Converts to bit/s
+		/// @ingroup modm_architecture_literals
+		/// @{
+		constexpr operator""_bps(T value);
+		constexpr operator""_kbps(T value);
+		constexpr operator""_Mbps(T value);
+		/// @}
+
+		/// Converts to Percent
+		/// @ingroup modm_architecture_literals
+		/// @{
+		constexpr operator""_pct(T value);
+		/// @}
+	}
+}
+
+#else
+
+#define MODM_LITERAL_DEFINITION(type, name, symbol) \
+namespace modm { \
+using MODM_CONCAT(name, _t) = type; \
+template<typename T> constexpr MODM_CONCAT(name, _t) symbol(T value) { return value; } \
+template<typename T> constexpr MODM_CONCAT(name, _t) MODM_CONCAT(k, symbol)(T value) { return value * 1'000ul; } \
+template<typename T> constexpr MODM_CONCAT(name, _t) MODM_CONCAT(M, symbol)(T value) { return value * 1'000'000ul; } \
+namespace literals { \
+	constexpr MODM_CONCAT(name, _t) MODM_CONCAT(operator""_ , symbol)(unsigned long long int value) { return symbol(value); } \
+	constexpr MODM_CONCAT(name, _t) MODM_CONCAT(operator""_k, symbol)(unsigned long long int value) { return MODM_CONCAT(k, symbol)(value); } \
+	constexpr MODM_CONCAT(name, _t) MODM_CONCAT(operator""_k, symbol)(long double value)            { return MODM_CONCAT(k, symbol)(value); } \
+	constexpr MODM_CONCAT(name, _t) MODM_CONCAT(operator""_M, symbol)(unsigned long long int value) { return MODM_CONCAT(M, symbol)(value); } \
+	constexpr MODM_CONCAT(name, _t) MODM_CONCAT(operator""_M, symbol)(long double value)            { return MODM_CONCAT(M, symbol)(value); } \
+}}
+
+MODM_LITERAL_DEFINITION(uint32_t, frequency, Hz)
+MODM_LITERAL_DEFINITION(uint32_t, baudrate, Bd)
+MODM_LITERAL_DEFINITION(uint32_t, bitrate, bps)
+
+namespace modm {
+enum class percent_t : uint16_t {};
+template<typename T> constexpr percent_t pct(T value) { return percent_t(uint16_t(value * 600ul)); }
+constexpr float pct2f(percent_t value) { return uint16_t(value) / 60'000.f; }
+namespace literals
+{
+	constexpr percent_t operator""_pct(long double value) { return pct(value); }
+	constexpr percent_t operator""_pct(unsigned long long int value) { return pct(value); }
+}
+}
```

Comment:
So… about the `enum class percent_t`: C++ prohibits using a class as a type for template parameters, only integrals and enums are allowed as non-type parameters and `float` is explicitly forbidden (how would you specialize a template on a float value?).

However, percent is not passed as a unit of one, like frequency/baudrate/bitrate is, so integers or floats must first be converted, requiring a strong type, and enum class is the only type that can be used as a template parameter type. I've added the `pct2f` (percent-to-float) function as a compromise to normalize the unit as required for tolerance checking.

In [C++20 using non-type template parameters](https://en.cppreference.com/w/cpp/language/template_parameters) is made possible with the `<=>` operator, and thus I plan to replace this hack with a proper class type, which provides implicit conversions from/to float.